app/main.py

```diff
@@ -1,34 +1,106 @@
-        pass
+        for row in range(start[0], end[0] + 1):
+            for column in range(start[1], end[1] + 1):
+                self.decks.append(Deck(row, column))
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if [deck.row, deck.column] == [row, column]:
+                return deck
+        return None
+
+    def fire(self, row: int, column: int) -> None:
+        deck = self.get_deck(row, column)
+        if deck:
+            deck.is_alive = False
+            if True not in [deck.is_alive for deck in self.decks]:
+                self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple[tuple[int, int]]]) -> None:
+        self.field = {
+            ship: Ship(ship[0], ship[1])
+            for ship in ships
+        }
+
+    def fire(self, location: tuple) -> str:
+        for ship in self.field:
+            if self.field[ship].get_deck(location[0], location[1]):
+                self.field[ship].fire(location[0], location[1])
+                if self.field[ship].is_drowned:
+                    return "Sunk!"
+                return "Hit!"
+        return "Miss!"
+
+    def print_field(self) -> None:
+        field = [["~"] * 10 for _ in range(10)]
+        for ship in self.field.values():
+            for deck in ship.decks:
+                if not deck.is_alive:
+                    symbol = "x" if ship.is_drowned else "*"
+                else:
+                    symbol = u"\u25A1"
+                field[deck.row][deck.column] = symbol
+        for row in field:
+            print(" ".join(row))
+
+    def _validate_field(self) -> None:
+        single_deck = double_deck = three_deck = four_deck = 0
+        for ship in self.field.values():
+            length = len(ship.decks)
+            if length == 1:
+                single_deck += 1
+            elif length == 2:
+                double_deck += 1
+            elif length == 3:
+                three_deck += 1
+            elif length == 4:
+                four_deck += 1
+
+        if not (single_deck == 4
+                and double_deck == 3
+                and three_deck == 2
+                and four_deck == 1):
+            raise ValueError("Incorrect number of ships")
```

Comment:
The `_validate_field` method checks for the correct number of ships but does not ensure that ships are placed within the bounds of the field. Additionally, it does not check if the ships are placed in a straight line, which is a common rule in Battleship games.